crates/sui-indexer/src/apis/indexer_api_v2.rs

```diff
@@ -291,22 +292,66 @@ impl IndexerApiServer for IndexerApiV2 {
     }
 
     async fn resolve_name_service_address(&self, name: String) -> RpcResult<Option<SuiAddress>> {
-        // TODO(manos): Implement new logic.
-        let domain = name
-            .parse::<Domain>()
-            .map_err(IndexerError::NameServiceError)?;
+        let domain: Domain = name.parse().map_err(IndexerError::NameServiceError)?;
+        let parent_domain = domain.parent();
 
+        // construct the record ids to lookup.
         let record_id = self.name_service_config.record_field_id(&domain);
+        let parent_record_id = self.name_service_config.record_field_id(&parent_domain);
+
+        // get latest timestamp to check expiration.
+        let current_timestamp = self.inner.get_latest_checkpoint()?.timestamp_ms;
+
+        // gather the requests to fetch in the multi_get_objs.
+        let mut requests = vec![record_id];
+
+        // we only want to fetch both the child and the parent if the domain is a subdomain.
+        if domain.is_subdomain() {
+            requests.push(parent_record_id);
+        }
 
-        let field_record_object = match self.inner.get_object_in_blocking_task(record_id).await? {
-            Some(o) => o,
-            None => return Ok(None),
+        // fetch both parent (if subdomain) and child records in a single get query.
+        // We do this as we do not know if the subdomain is a node or leaf record.
+        let mut domains: Vec<_> = self
+            .inner
+            .multi_get_objects_in_blocking_task(requests)
+            .await?
+            .into_iter()
+            .map(|o| sui_types::object::Object::try_from(o).ok())
+            .collect();
+
+        let name_record = match name_record_from_option_obj_vec(&mut domains, &record_id) {
+            Ok(Some(name_record)) => name_record,
+            Ok(None) => return Ok(None),
+            Err(e) => return Err(IndexerError::NameServiceError(e).into()),
```

Comment:
So it looks like we only ever fetch 2 things at most, the record and possibly the parent record. iirc the return order of mult-fetches are in the order of their arguments so you could destructure the return'd vector and attempt to convert to the name record in-place vs using this `name_record_from_option_obj_vec` which has a fairly complicated contract.

Reply:
@bmwill when I looked in the implementation for multi_get_objects (`multi_get_objects_impl`), it looks like it converts into an `object_id IN ()` db query which, after looking it up, seems to not guarantee the order of the results. I was worried that destructuring them in place and doing the checks would not look really good either. 

Thoughts?

Reply:
@bmwill I've addressed the comment and moved the logic in the calling function, without any external helpers.